multidict/_multidict_c.c

```diff
@@ -0,0 +1,1617 @@
+    if (_MultiDict_Check(arg) && (kwds == NULL || kwds == Py_None)) {
+        if (MultiDict_CheckExact(arg) || CIMultiDict_CheckExact(arg)) {
+            impl = ((MultiDictObject*)arg)->impl;
+        } else if (MultiDictProxy_CheckExact(arg) || CIMultiDictProxy_CheckExact(arg)) {
+            impl = ((MultiDictProxyObject*)arg)->md->impl;
+        }
+
+        if (do_add) {
+            return _multidict_append_items(self, impl);
+        }
+
+        return _multidict_update_items(self, impl);
+    }
+
+    if (PyObject_HasAttrString(arg, "items")) {
+        if (_MultiDict_Check(arg)) {
+            arg_items = multidict_items((MultiDictObject*)arg);
+        } else {
+            arg_items = PyMapping_Items(arg);
+        }
+        if (arg_items == NULL) {
+            return -1;
+        }
+    } else {
+        arg_items = arg;
+        Py_INCREF(arg_items);
+    }
+
+    if (kwds && kwds != Py_None) {
+        kwds_items = PyDict_Items(kwds);
+        err = _multidict_list_extend(arg_items, kwds_items);
+        Py_DECREF(kwds_items);
+        if (err < 0) {
+            Py_DECREF(arg_items);
+            return -1;
+        }
+    }
+
+    if (do_add) {
+        err = _multidict_append_items_seq(self, arg_items, name);
+    } else {
+        err =  pair_list_update_from_seq(self->impl, arg_items);
+    }
+
+    Py_DECREF(arg_items);
+
+    return err;
+}
+
+static INLINE int
+_multidict_extend_with_kwds(MultiDictObject *self, PyObject *kwds,
+                                 const char *name, int do_add)
+{
+    PyObject *arg = NULL;
+
+    int err = 0;
+
+    arg = PyDict_Items(kwds);
+    if (do_add) {
+        err = _multidict_append_items_seq(self, arg, name);
+    } else {
+        err = pair_list_update_from_seq(self->impl, arg);
+    }
+
+    Py_DECREF(arg);
+    return err;
+}
+
+static int
+_multidict_extend(MultiDictObject *self, PyObject *args, PyObject *kwds,
+                  const char *name, int do_add)
+{
+    PyObject *arg = NULL;
+
+    if (args && PyObject_Length(args) > 1)  {
+        PyErr_Format(
+            PyExc_TypeError,
+            "%s takes at most 1 positional argument (%zd given)",
+            name, PyObject_Length(args), NULL
+        );
+        return -1;
+    }
+
+    if (args && PyObject_Length(args) > 0) {
+        if (!PyArg_UnpackTuple(args, name, 0, 1, &arg)) {
+            return -1;
+        }
+        if (_multidict_extend_with_args(self, arg, kwds, name, do_add) < 0) {
+            return -1;
+        }
+    } else if (kwds && PyObject_Length(kwds) > 0) {
+        if (_multidict_extend_with_kwds(self, kwds, name, do_add) < 0) {
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static INLINE PyObject *
+_multidict_copy(MultiDictObject *self, PyTypeObject *multidict_tp_object)
+{
+    MultiDictObject *new_multidict = NULL;
+
+    PyObject *arg_items = NULL,
+             *items     = NULL;
+
+    new_multidict = PyObject_GC_New(MultiDictObject, multidict_tp_object);
```

Comment:
Forgot to call `PyObject_GC_TRACK()` maybe?
Or just `new_multidict = multidict_tp_object->tp_new()`?

Reply:
Switched to PyType_GenericNew.